Add Repeating Basis Functions

RensDimmendaal opened the pull request:
I worked on this a while ago in PR #147, but I started from a fresh branch because I decided to limit the scope only to repeating basis functions (#20), excluding the spanning basis functions (#29).

Feedback adapted so far:
- Write a transformer when X contains only one column
- Write a wrapper which selects one column, when X has more than one column
- Added basic tests
- Added docstring

To Do:
- Write documentation

Could someone review the code? 
I'll work on the documentation in the meantime as well.

Tagging people involved in PR #147: @kayhoogland @koaning @MBrouns

koaning commented on sklego/preprocessing.py:
```diff
@@ -609,3 +610,133 @@ def _check_X_for_type(X):
+    This is a transformer for features with some form of circularity. 
+    E.g. for days of the week you might face the problem that, conceptually, day 7 is as 
+    close to day 6 as it is to day 1. While numerically their distance is different.
+    This transformer solves that problem.
```
`solve` might be too strong a word, maybe `remedy`?

koaning commented on sklego/preprocessing.py:
```diff
@@ -609,3 +610,133 @@ def _check_X_for_type(X):
+    E.g. for days of the week you might face the problem that, conceptually, day 7 is as 
+    close to day 6 as it is to day 1. While numerically their distance is different.
+    This transformer solves that problem.
+    The transformer selects a column and transforms it with a give number of repeating (radial) basis functions.
```
"with a give" -> spelling

koaning commented on sklego/preprocessing.py:
```diff
@@ -609,3 +610,133 @@ def _check_X_for_type(X):
+    :param n_periods: number of basis functions to create, i.e., the number of columns that
+        will exit the transformer.
+
+    :type floor: float or 'min', default='min'
```
I am a bit unsure what you mean with the floor/ceil parameters here. The RBF spans between 0-1 no? 

![image](https://user-images.githubusercontent.com/1019791/62199297-841b9a80-b383-11e9-9a89-256d84958763.png)

RensDimmendaal replied:
Thanks. Two actionable points of feedback come out of this:
1. Ensure the curves always range between 0 and 1. Now this is note the case, but that would be neater.
2. Clearly label what ceil/floor do. What they do is scale the input data. Imagine you want to make repeating basis functions on a column with hour of the day. But your training data does not include data points between midnight and 3am. You might still want to have your repeating basis functions be split on the 0-24 range rather than the 3-24 range. Explicitly saying what the minimum and maximum of your input range fixes this.

Will take them on board next week

RensDimmendaal replied:
I've renamed floor and ceil to one parameter `input_range` which accepts a tuple. I hope it now better reflects what it does: it captures at what values the input data starts over again. E.g. for days of the week you'd provide (1,7)

koaning commented:
I've taken a few minutes to have a glance, will try to find some more time to look at the code later.

RensDimmendaal commented:
* Adopted the docstring suggestions
* Rename the floor/ceil parameters to input_range to clarify what it does
* Added documentation as a notebook

Not sure why travis fails, seems to me like the issue is with other notebooks which i have not touched [link](https://travis-ci.com/koaning/scikit-lego/jobs/224227006).

koaning commented:
So there's a bug in a part of the code that has to do with the `grouped-estimator`. If that part of the library is causing this then I'll still gladly merge the work here.

koaning commented:
![image](https://user-images.githubusercontent.com/1019791/62854236-ff1f7200-bcee-11e9-97cf-bc3b8636d81b.png)

but this seems to be formatting on your side?

koaning commented:
But this you may safely ignore, this is something that I am picking up with @pim-hoeven. 

![image](https://user-images.githubusercontent.com/1019791/62854276-25dda880-bcef-11e9-8308-46616595d5d3.png)

RensDimmendaal commented:
Thanks. My bad on the formatting issue. Fixed it now.

![image](https://user-images.githubusercontent.com/9828683/62855013-4dce0b80-bcf1-11e9-85f9-ab20eafe9cd1.png)

If that other issue can be ignored then I think it is ready.

koaning commented:
Great. I'll have a look at the end of the day.

koaning commented:
![image](https://user-images.githubusercontent.com/1019791/62973129-c6c68380-be15-11e9-9169-be05c20c1aa1.png)

Great that you took the effort of moving the preprocessing documentation into a notebook but is there a reason why only your code is evaluated?

koaning commented:
I am playing with the notebook and i see something odd. 

![image](https://user-images.githubusercontent.com/1019791/62973484-726fd380-be16-11e9-9f5a-1b732e78d65e.png)

![image](https://user-images.githubusercontent.com/1019791/62973494-78fe4b00-be16-11e9-9d46-0d964e9116c5.png)


When remainder should be dropping ... it isn't?

koaning commented:
I am having worries about the remainder behaviour. To me it feels like the base behaviour should be that it always drops the column. In my mind it would be used similar to this: 

![image](https://user-images.githubusercontent.com/1019791/62973783-18bbd900-be17-11e9-8a2f-e7373fd0226e.png)

You would have a feature union split the steps and pass the correct column to the model. I think having passthrough behaviour can still be valid but I would argue that it goes against a base behaviour. 

@MBrouns thoughts?

RensDimmendaal commented:
**Docs**:
Had to do a slight rewrite of the code snippet you screenshotted above, but I've now converted all snippets in the preprocessing docs to working code.

**Regarding faulty "drop" behavior**:
Shame on me, that's a bug. I've fixed it and added a test for it now as well.

**Default behavior "drop"/"passthrough"**:
That's fair. It makes more sense to have "drop" as the default. That's also the default of the ColumnTransformer which I use to accomplish the behavior.

MBrouns commented:
regarding the drop / passthrough I think we should go for the principal of least surprise and to me it feels like the transformer is most like the one hot encoder. The OHE keeps all the non-categorical columns and simply adds the encoded columns so for me it would make most sense to have the same behaviour here. 

On the other hand, the `categorical_features` attribute of the OHE is getting deprecated and users are told to use a `ColumnTransformer` instead, so I don't know..

koaning commented:
Oh yeah. I forgot to click the merge button. @RensDimmendaal well done!

RensDimmendaal commented:
Whoo! :-)
